src/rust/engine/nailgun/src/client.rs

```diff
@@ -0,0 +1,152 @@
+// Copyright 2020 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+#![deny(warnings)]
+// Enable all clippy lints except for many of the pedantic ones. It's a shame this needs to be copied and pasted across crates, but there doesn't appear to be a way to include inner attributes from a common source.
+#![deny(
+  clippy::all,
+  clippy::default_trait_access,
+  clippy::expl_impl_clone_on_copy,
+  clippy::if_not_else,
+  clippy::needless_continue,
+  clippy::unseparated_literal_suffix,
+// TODO: Falsely triggers for async/await:
+//   see https://github.com/rust-lang/rust-clippy/issues/5360
+// clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+use nails::execution::{stream_for, ChildInput, ChildOutput, ExitCode};
+use nails::Config;
+use tokio::net::TcpStream;
+
+use std::io;
+use std::net::Ipv4Addr;
+use tokio::io::AsyncWriteExt;
+
+use futures::channel::{mpsc, oneshot};
+use futures::{future, FutureExt, SinkExt, Stream, StreamExt};
+use log::debug;
+
+pub enum NailgunClientError {
+  PreConnect(String),
+  PostConnect(String),
+  ExplicitQuit,
+}
+
+async fn handle_client_output(
+  mut stdio_read: impl Stream<Item = ChildOutput> + Unpin,
+) -> Result<(), io::Error> {
+  let mut stdout = tokio::io::stdout();
+  let mut stderr = tokio::io::stderr();
+  while let Some(output) = stdio_read.next().await {
+    match output {
+      ChildOutput::Stdout(bytes) => stdout.write_all(&bytes).await?,
+      ChildOutput::Stderr(bytes) => stderr.write_all(&bytes).await?,
+      ChildOutput::Exit(_) => {
+        // NB: We ignore exit here and allow the main thread to handle exiting.
```

Comment:
Where does the main exiting happen? If might be nice to just say what we expect the main thread to be doing in order to ensure exiting -- I'm worried the comment could go stale is all.